# Create alternate reference genomes

Call SNPs from whole genome resequencing data of the parents to create alternate reference genomes for better mapping of the WGBS and to use the SNPs to pull out parental alleles from the WGBS. 

---

**Methods:** Whole genome sequencing data were quality checked using fastqc v0.11.8 ([Andrews 2010](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)). Data were aligned to the reference genome Bter_1.0 (Refseq accession no. GCF_000214255.1, [Sadd et al. 2015](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0623-3)) using Bowtie2 v2.3.5.1 ([Langmead and Salzberg 2013](https://www.nature.com/articles/nmeth.1923)). Alignments were deduplicated using Picard v2.23.8 ([Broad Institute 2018](https://broadinstitute.github.io/picard/)) and reads were realigned around indels using GATK3 v3.8 ([McKenna et al. 2010](https://genome.cshlp.org/content/20/9/1297)). SNPs were then called from each parent separately using Freebayes v0.9.21.7 ([Garrison and Marth 2012](https://arxiv.org/abs/1207.3907)) which takes ploidy into account. SNPs were filtered using VCFtools v0.1.16 ([Denecek et al. 2011](https://academic.oup.com/bioinformatics/article/27/15/2156/402296)) with the following options --max-alleles 2 --minQ 20 --min-meanDP 10. Bedtools intsect (REF XXX) was then used to identify SNPs unique to the mother or father from each cross to enable the identification of parental alleles in the offspring.


**Notes:** Fastqc output looks great! No need for trimming, wowza! After quality filtering we still have just under 1million SNPs per sample, this is great as the number will greatly reduce when I filter for informative SNPs only. This reduces to between 100,000 - 400,000 'informative' SNPS, so much better than the 40,000 from the WGBS! Woo!!