---
title: "MethlKit_Male_Queen"
author: "Hollie Marshall"
date: "11/30/2018"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Differential Methylation Analysis

This is the script for a differential methylation analysis between reproductive workers, males and queen bumblebees. The data is whole genome bisulfite sequencing data, this has been alined to the reference genome Bter_1.0 and deduplicated using Bismark. The resulting bam files have also been sorted and indexed using samtools. The worker samples are pools of 3 individuals per colony but the queen/male samples are the partents of each colony. There are 4 colonies total.

```{r Admin}
setwd("./")
library(methylKit)
library(grid)
library(readr)
library(ggplot2)
```

The following commands reads a sorted BAM file and creates a methylRaw object for CpG methylation. It also writes out .txt files for each input making future use easier as it takes much longer to read in a BAM compared to a TXT. (Do this for each analysis, as need the 'treatment' variable to match later).
```{r Make Convenient Input Files}
file.list <- list("m08_trimmed_deduplicated_sorted.bam","m19_trimmed_deduplicated_sorted.bam","m23_trimmed_deduplicated_sorted.bam","m37_trimmed_deduplicated_sorted.bam",
"q08_trimmed_deduplicated_sorted.bam","q19_trimmed_deduplicated_sorted.bam", "q23_trimmed_deduplicated_sorted.bam", "q37_trimmed_deduplicated_sorted.bam")

raw_data <- processBismarkAln(file.list,
                        sample.id = list("m08","m19","m23","m37","q08","q19", "q23", "q37"),
                         treatment = c(1,1,1,1,0,0,0,0),
                         assembly="bter_1.0", 
                         read.context="CpG",
                         save.context = "CpG",
                         save.folder=getwd(),
                         save.db = TRUE)
```


After the above has been run you can use the below code each time to read in much smaller files based on the original .bams, this saves time and memory for future analyses.
```{r Input Files}
file.list <- list("m08_CpG.txt", "m19_CpG.txt", "m23_CpG.txt", "m37_CpG.txt", "q08_CpG.txt", "q19_CpG.txt", "q23_CpG.txt", "q37_CpG.txt")

raw_data <- methRead(file.list,
              sample.id = list("m08","m19","m23","m37","q08","q19", "q23", "q37"),
              treatment = c(1,1,1,1,0,0,0,0),
              assembly="bter_1.0", 
              context="CpG")
```

Get some stats on the objects and write out as one .pdf file with multiple plots in it. (Need to play with this, the following commands give the plot for just the second dataset in the methylobject list).
```{r Initial Look at Data}
pdf("initial_look.pdf") 
getMethylationStats(raw_data[[2]],plot=TRUE,both.strands=FALSE)
dev.off()

pdf("initial_look2.pdf") 
getCoverageStats(raw_data[[2]],plot=TRUE,both.strands=FALSE)
dev.off()
```

Filter data by coverage; must have at least 10 reads per CpG and not exceed the 99.9th percentile (PCR duplicates), also tested with only 4 reads and similar results in terms of the number of diff methylated CpGs.
```{r Filtering for Coverage}
filtered_data <- filterByCoverage(raw_data,lo.count=10,lo.perc=NULL,
                                      hi.count=NULL,hi.perc=99.9)
```

Select only CpGs found in all 6 samples for analysis (destand =T can be used for CpG methylation to use both strands and incresae coverage, this should not be used for non-CpG though). You can decrease the minimum number of samples a particular CpG occurs in by using 'min.per.group =2L'.
```{r Combine Data}
meth_all_data <- unite(filtered_data, destrand=TRUE)
```

http://www.stat.purdue.edu/~cheng70/code.html to select which sites are methylated and then filter the final dataframe on these rather than an arbituary cut-off, i.e. at least one sample should contain methylation per position in order for it to be considered. This protocol uses a binomial model to decide if a site is methylated. (TO DO: Put below code into nice little loop/apply.)
```{r MSC Methylation Call Filtering}
df_meth_all <- getData(meth_all_data)
df_meth_all$rownums <- row.names(df_meth_all)

source("/scratch/monoallelic/hm257/repro_methylation/bams/meth_calling_R_code/MSC.R")
source("/scratch/monoallelic/hm257/repro_methylation/bams/meth_calling_R_code/rateestimate.R")

samp1 <- df_meth_all[,5:6]
colnames(samp1) <- c("CT","Ccount")
MSCount_samp1 <- MSC(samp1, 1e-08)
MSCresult1 <- MSCount_samp1$MSC
write.csv(MSCresult1, file="MSC_meth_calls_m08.csv")
pi1 <- MSCount_samp1$pi
MSCrate1 <- rateestimate(MSCresult1,pi1)
rate1 <- as.data.frame(MSCrate1)
write.csv(rate1, file="MSC_rate_estimates_m08.csv")
MSCresult1$row_nums <- row.names(MSCresult1)
MSCresult1_meth <- subset(MSCresult1, status == "methylated")

samp2 <- df_meth_all[,8:9]
colnames(samp2) <- c("CT","Ccount")
MSCount_samp2 <- MSC(samp2, 1e-08)
MSCresult2 <- MSCount_samp2$MSC
write.csv(MSCresult2, file="MSC_meth_calls_m19.csv")
pi2 <- MSCount_samp2$pi
MSCrate2 <- rateestimate(MSCresult2,pi2)
rate2 <- as.data.frame(MSCrate2)
write.csv(rate2, file="MSC_rate_estimates_m19.csv")
MSCresult2$row_nums <- row.names(MSCresult2)
MSCresult2_meth <- subset(MSCresult2, status == "methylated")

samp3 <- df_meth_all[,11:12]
colnames(samp3) <- c("CT","Ccount")
MSCount_samp3 <- MSC(samp3, 1e-08)
MSCresult3 <- MSCount_samp3$MSC
write.csv(MSCresult3, file="MSC_meth_calls_m23.csv")
pi3 <- MSCount_samp3$pi
MSCrate3 <- rateestimate(MSCresult3,pi3)
rate3 <- as.data.frame(MSCrate3)
write.csv(rate3, file="MSC_rate_estimates_m23.csv")
MSCresult3$row_nums <- row.names(MSCresult3)
MSCresult3_meth <- subset(MSCresult3, status == "methylated")

samp4 <- df_meth_all[,14:15]
colnames(samp4) <- c("CT","Ccount")
MSCount_samp4 <- MSC(samp4, 1e-08)
MSCresult4 <- MSCount_samp4$MSC
write.csv(MSCresult4, file="MSC_meth_calls_m37.csv")
pi4 <- MSCount_samp4$pi
MSCrate4 <- rateestimate(MSCresult4,pi4)
rate4 <- as.data.frame(MSCrate4)
write.csv(rate4, file="MSC_rate_estimates_m37.csv")
MSCresult4$row_nums <- row.names(MSCresult4)
MSCresult4_meth <- subset(MSCresult4, status == "methylated")

samp5 <- df_meth_all[,17:18]
colnames(samp5) <- c("CT","Ccount")
MSCount_samp5 <- MSC(samp5, 1e-08)
MSCresult5 <- MSCount_samp5$MSC
write.csv(MSCresult5, file="MSC_meth_calls_q08.csv")
pi5 <- MSCount_samp5$pi
MSCrate5 <- rateestimate(MSCresult5,pi5)
rate5 <- as.data.frame(MSCrate5)
write.csv(rate5, file="MSC_rate_estimates_q08.csv")
MSCresult5$row_nums <- row.names(MSCresult5)
MSCresult5_meth <- subset(MSCresult5, status == "methylated")

samp6 <- df_meth_all[,20:21]
colnames(samp6) <- c("CT","Ccount")
MSCount_samp6 <- MSC(samp6, 1e-08)
MSCresult6 <- MSCount_samp6$MSC
write.csv(MSCresult6, file="MSC_meth_calls_q19.csv")
pi6 <- MSCount_samp6$pi
MSCrate6 <- rateestimate(MSCresult6,pi6)
rate6 <- as.data.frame(MSCrate6)
write.csv(rate6, file="MSC_rate_estimates_q19.csv")
MSCresult6$row_nums <- row.names(MSCresult6)
MSCresult6_meth <- subset(MSCresult6, status == "methylated")

samp7 <- df_meth_all[,23:24]
colnames(samp7) <- c("CT","Ccount")
MSCount_samp7 <- MSC(samp7, 1e-08)
MSCresult7 <- MSCount_samp7$MSC
write.csv(MSCresult7, file="MSC_meth_calls_q23.csv")
pi7 <- MSCount_samp7$pi
MSCrate7 <- rateestimate(MSCresult7,pi7)
rate7 <- as.data.frame(MSCrate7)
write.csv(rate7, file="MSC_rate_estimates_q23.csv")
MSCresult7$row_nums <- row.names(MSCresult7)
MSCresult7_meth <- subset(MSCresult7, status == "methylated")

samp8 <- df_meth_all[,26:27]
colnames(samp8) <- c("CT","Ccount")
MSCount_samp8 <- MSC(samp8, 1e-08)
MSCresult8 <- MSCount_samp8$MSC
write.csv(MSCresult8, file="MSC_meth_calls_q37.csv")
pi8 <- MSCount_samp8$pi
MSCrate8 <- rateestimate(MSCresult8,pi8)
rate8 <- as.data.frame(MSCrate8)
write.csv(rate8, file="MSC_rate_estimates_q37.csv")
MSCresult8$row_nums <- row.names(MSCresult8)
MSCresult8_meth <- subset(MSCresult8, status == "methylated")


all_data <- rbind(MSCresult1_meth, MSCresult2_meth, MSCresult3_meth, MSCresult4_meth, MSCresult5_meth, MSCresult6_meth, MSCresult7_meth, MSCresult8_meth)

meth_positions <- as.vector(as.numeric(unique(all_data$row_nums))) 

subset_methBase <- select (meth_all_data, meth_positions)
```


Check sample correlations.
```{r Correlations}
pdf("correlation_male_queen.pdf")
getCorrelation(subset_methBase,plot=TRUE)
dev.off()
```

Look at a hierarchical cluster of the samples, to make a nicer image use the output in the second line of code to feed into a dendogram program.
```{r Clustering}
pdf("sample_cluster_male_queen.pdf")
clusterSamples(subset_methBase, dist="correlation", method="ward", plot=TRUE)
dev.off()

hc <- clusterSamples(subset_methBase, dist="correlation", method="ward", plot=FALSE)
```

Do a PCA, the screeplot tells you which PCs store the most varience.
```{r PCA and Scree Plots}
pdf("screeplot_male_queen.pdf")
PCASamples(subset_methBase, screeplot=TRUE)
dev.off()

pdf("pca_plot_male_queen.pdf")
PCASamples(subset_methBase)
dev.off()
```


Find differentially methylated regions with a min percentage diff of 10 and q-val of 0.05. It uses a logistic regression and the SLIM method for correction for multiple testing. Can change the number of cores if using on a HPC. Family referes to the way the crosses were carried out, i.e. m08 is the brother of q19 etc.
```{r Differential Methylation Anslysis}
covariates <- data.frame(subspecies=c("audax","dalmatinus", "audax", "dalmatinus", "dalmatinus", "audax", "dalmatinus", "audax"), family=c("1","2","3","4","2","1","4","3"))

diff_meth <- calculateDiffMeth(subset_methBase, covariates=covariates, mc.cores = 3)
write.csv(diff_meth, file="All_tested_meth_sites_MSCfilter_worker_queen.csv")

diff_meth_10 <- getMethylDiff(diff_meth, difference=10, qvalue=0.05)
write.csv(diff_meth_10, file="DMRs_min10percentDiff_qval0.05_MSCfilter_worker_queen.csv")
```



# Step 2. Making small inputs to make the graphs (PCA etc.) nicer

Use the set of positions generated above which have been filtered on coverage and have at least one sample showing methylation per base.
```{r Get Data for subset}
methBase_ob <- getData(subset_methBase)
```

Write out these positions for later use.
```{r Write out Positions}
write.table(methBase_ob, file="objectmethbase.txt", quote=F, 
          row.names = F, sep = '\t')
```

Make new input files based on the information in the above methBase object. Read in data if needed, add a chrBase column, and remove a start/end information column. Also change the '+' in strand to 'F', there is no reverse strand as the original data above is already destranded. 
```{r}
objectmethbase <- read_delim("objectmethbase.txt", 
                             "\t", escape_double = FALSE, trim_ws = TRUE)

objectmethbase$chrBase <- paste(objectmethbase$chr, ".", objectmethbase$start, sep="")
objectmethbase <- objectmethbase[,-3]
objectmethbase$strand <- gsub(pattern = "\\+", replace = "F", objectmethbase$strand)
```

Make the new files, based on the correct columns of coverage/numC/numT data.
"m08","m19","m23","m37","q08","q19", "q23", "q37"
```{r}
m08 <- objectmethbase[,c("chrBase","chr","start","strand","coverage1","numCs1","numTs1")]
m19 <- objectmethbase[,c("chrBase","chr","start","strand","coverage2","numCs2","numTs2")]
m23 <- objectmethbase[,c("chrBase","chr","start","strand","coverage3","numCs3","numTs3")]
m37 <- objectmethbase[,c("chrBase","chr","start","strand","coverage4","numCs4","numTs4")]
q08 <- objectmethbase[,c("chrBase","chr","start","strand","coverage5","numCs5","numTs5")]
q19 <- objectmethbase[,c("chrBase","chr","start","strand","coverage6","numCs6","numTs6")]
q23 <- objectmethbase[,c("chrBase","chr","start","strand","coverage7","numCs7","numTs7")]
q37 <- objectmethbase[,c("chrBase","chr","start","strand","coverage8","numCs8","numTs8")]
```

Make a list of these objects. Rename to match methylKit input. Add frequency columns and remove the 'numC/numT' columns. Rounding to two decimals produces frequencies that add to 100 in every case. Make final output files.
```{r}
all_files <- list(m08, m19, m23, m37, q08, q19, q23, q37)

for(i in seq_along(all_files)){
  colnames(all_files[[i]])[c(3,5,6,7)] <- c("base","coverage","numCs","numTs")
  all_files[[i]]$freqC <- round((all_files[[i]]$numCs/all_files[[i]]$coverage)*100,
                                digits=2)
  all_files[[i]]$freqT <- round((all_files[[i]]$numTs/all_files[[i]]$coverage)*100,
                                digits=2)
  all_files[[i]] <- all_files[[i]][-c(6,7)]
  myfile <- file.path("./", paste0(i,"_","subsetted_final.txt"))
  write.table(all_files[[i]], file=myfile, quote=F, sep="\t", row.names=F)
}
```
